--- dwm.c	2011-05-29 11:54:32.275200225 -0400
+++ dwm.c	2011-05-29 12:00:08.206713665 -0400
@@ -168,6 +168,7 @@
 static void configurenotify(XEvent *e);
 static void configurerequest(XEvent *e);
 static Monitor *createmon(void);
+static void cycle(const Arg *arg);
 static void destroynotify(XEvent *e);
 static void detach(Client *c);
 static void detachstack(Client *c);
@@ -213,10 +214,12 @@
 static void setlayout(const Arg *arg);
 static void setmfact(const Arg *arg);
 static void setup(void);
+static int shifttag(int dist);
 static void showhide(Client *c);
 static void sigchld(int unused);
 static void spawn(const Arg *arg);
 static void tag(const Arg *arg);
+static void tagcycle(const Arg *arg);
 static void tagmon(const Arg *arg);
 static int textnw(const char *text, unsigned int len);
 static void tile(Monitor *);
@@ -2062,3 +2065,42 @@
 	XCloseDisplay(dpy);
 	return 0;
 }
+
+static int
+shifttag(int dist)
+{
+        int i, curtags;
+        int seltag = 0;
+        int numtags = LENGTH(tags);
+
+        curtags = selmon->tagset[selmon->seltags];
+        for(i = 0; i < LENGTH(tags); i++) {
+                if((curtags & (1 << i)) != 0) {
+                        seltag = i;
+                        break;
+                }
+        }
+
+        seltag += dist;
+        if(seltag < 0)
+                seltag = numtags - (-seltag) % numtags;
+        else
+                seltag %= numtags;
+
+        return 1 << seltag;
+}
+
+static void
+cycle(const Arg *arg)
+{
+        const Arg a = { .i = shifttag(arg->i) };
+        view(&a);
+}
+
+static void
+tagcycle(const Arg *arg)
+{
+        const Arg a = { .i = shifttag(arg->i) };
+        tag(&a);
+        view(&a);
+}
