--- dwm.c.orig	2011-05-29 19:23:32.982583221 -0400
+++ dwm.c	2011-05-29 19:33:15.399427997 -0400
@@ -118,6 +118,7 @@
 } Key;
 
 typedef struct {
+        const char *icon;
 	const char *symbol;
 	void (*arrange)(Monitor *);
 } Layout;
@@ -176,6 +177,7 @@
 static Monitor *dirtomon(int dir);
 static void drawbar(Monitor *m);
 static void drawbars(void);
+static void drawicon(const char *file, unsigned long col[ColLast]);
 static void drawsquare(Bool filled, Bool empty, Bool invert, unsigned long col[ColLast]);
 static void drawtext(const char *text, unsigned long col[ColLast], Bool invert);
 static void enternotify(XEvent *e);
@@ -705,8 +707,17 @@
 		           occ & 1 << i, urg & 1 << i, col);
 		dc.x += dc.w;
 	}
-	dc.w = blw = TEXTW(m->ltsymbol);
-	drawtext(m->ltsymbol, dc.norm, False);
+        blw = TEXTW(m->ltsymbol);
+
+        if(useicons) {
+                dc.w = dc.h;
+                drawicon(m->lt[m->sellt]->icon, dc.colors[0]);
+        }
+        else {
+                dc.w = blw;
+                drawtext(m->ltsymbol, dc.colors[0], False);
+        }
+ 
 	dc.x += dc.w;
 	x = dc.x;
 	if(m == selmon) { /* status is only drawn on selected monitor */
@@ -743,6 +754,20 @@
 }
 
 void
+drawicon(const char *file, unsigned long col[ColLast]) {
+        Pixmap icon;
+        int px, py;
+        unsigned int pw, ph;
+        icon = XCreatePixmap(dpy, dc.drawable, dc.h, dc.h, 1);
+        XReadBitmapFile(dpy, dc.drawable, file, &pw, &ph, &icon, &px, &py);
+        XSetForeground(dpy, dc.gc, col[ColBG]);
+        XFillRectangle(dpy, dc.drawable, dc.gc, dc.x, 0, dc.h, dc.h);
+        XSetForeground(dpy, dc.gc, col[ColFG]);
+        XSetBackground(dpy, dc.gc, col[ColBG]);
+        XCopyPlane(dpy, icon, dc.drawable, dc.gc, px, py, pw, ph, dc.x, dc.h/2-ph/2+py, 1);
+}
+
+void
 drawsquare(Bool filled, Bool empty, Bool invert, unsigned long col[ColLast]) {
 	int x;
 	XGCValues gcv;
