diff --git a/dwm.c b/dwm.c
index 803cd10..701d3d5 100644
--- a/dwm.c
+++ b/dwm.c
@@ -774,17 +774,13 @@ drawbar(Monitor *m) {
 	drawtext(m->ltsymbol, dc.colors[0], True);
 	dc.x += dc.w;
 	x = dc.x;
-	if(m == selmon) { /* status is only drawn on selected monitor */
-		dc.w = TEXTW(stext);
-		dc.x = m->ww - dc.w;
-		if(dc.x < x) {
-			dc.x = x;
-			dc.w = m->ww - x;
-		}
-		drawcoloredtext(stext);
-	}
-	else
-		dc.x = m->ww;
+        dc.w = TEXTW(stext);
+        dc.x = m->ww - dc.w;
+        if(dc.x < x) {
+                dc.x = x;
+                dc.w = m->ww - x;
+        }
+        drawcoloredtext(stext);
 	if((dc.w = dc.x - x) > bh) {
 		dc.x = x;
 		if(m->sel) {
@@ -2119,7 +2119,7 @@ void
 updatestatus(void) {
 	if(!gettextprop(root, XA_WM_NAME, stext, sizeof(stext)))
 		strcpy(stext, "dwm-"VERSION);
-	drawbar(selmon);
+	drawbars();
 }
 
 void
